### ISSUE TITLE
High CPU usage on GKE

### ISSUE THREAD

**Issue by rachlies** [SocketCluster/socketcluster#296]
I am using google container engine to create cluster with 3 vm instances having following configs.

1. Machine type	n1-standard-1 (1 vCPU, 3.75 GB memory)
2. Total cores	3 vCPUs
3. Total memory	11.25 GB
4. Disk size in GB (per node)	100
5. Local SSD disks (per node)	0

I have socketcluster(3 replicas pods), 1 scc-state pod, and 1 scc-broker pod, running in those vms. Those three socketcluster pods are evenly distributed among the nodes. Now for performance testing I tried testing it with 50 concurrent clients only and each client sending 100 msg/second simultaneously. 

But my CPU usage in each VM nodes was reaching 100% utilization as soon as I started the script. And also I am not able to understand that why the memory usage is 7.43 Gi even if it is idle.

**Comment by jondubois** [comment]
@rachlies Are you using channels? If so, are the published messages being read/consumed by every client? If so then 100 msg/second * 50 clients is 5000 inbound messages per second in total... Now If you have 50 clients each consuming 5000 messages per second then that's 250K outbound messages which have to be sent out to clients every second.

You can tell SC to batch messages together at an interval by setting a `pubSubBatchDuration` option (see http://socketcluster.io/#!/docs/api-socketcluster); this can give a significant speedup in certain scenarios like described above.

The memory consumption looks very high. I'm not sure what the cause is. I did not experience this issue on Amazon EC2.

**Comment by happilymarrieddad** [comment]
We have a lot of clients on our systems and so far we haven't had any memory or CPU usage issues. It could be some rogue code or something causing the issue. Are you sure it's socketcluster and not some async code that you wrote?

**Comment by rachlies** [comment]
@jondubois,  what is the message limit I should expect with the configuration that I mentioned.
@happilymarrieddad I just modified the worker.js according to the examples shown in the Docs. and then build a docker image of it to use it in the GKE. 

```
scServer.on('connection', function (socket) {
    console.log('User connected: ', count++);
    socket.on('chat', function (data) {
      var msg = '', channel_name = '';
      if(typeof data == 'object') {
        msg = data.msg;
        channel_name = data.channel_name;  
      }
      if(msg == undefined || msg == '')
        msg = 'hello';
      if(channel_name ==  undefined || channel_name == '')
        channel_name = 'broadcast';
      scServer.global.publish(channel_name, data);
      console.log("channel_name: ", channel_name);
      console.log('Chat:', data);
    });

    socket.on('disconnect', function () {
      console.log('User disconnected');
    });
  });
```
Following is the result shown when I run `top` in one of the node. As u can see aroung 50% memory is used in cached memory.  
top - 09:49:28 up 1 day, 22:44,  1 user,  load average: 0.00, 0.00, 0.00
Tasks: 114 total,   1 running, 113 sleeping,   0 stopped,   0 zombie
%Cpu(s):  1.3 us,  1.0 sy,  0.0 ni, 97.7 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
KiB Mem:   3794460 total,  2840072 used,   954388 free,   266064 buffers
KiB Swap:        0 total,        0 used,        0 free.  1803020 cached Mem

**Comment by toredash** [comment]
@jondubois have you have any experience with `pubSubBatchDuration` and the affects of increasing this value? I'm only interested in the affects it could have on the servers performance.

I looked at the code in `ncom`, and if I understand it correctly, by increasing this value one could reduce the number of websocket frames used to sent messages to subscribers, given that the messages was emitted at almost the same time.

I looked at some stats for one of our broadcast-type channels, and from the timeline of messages that was sent out, quite a few of them was emitted at the same time. For instance we have 54 messages sent out at this specific time: 2017-11-19T02:53:56.310Z

I'm finding it hard to measure the performance implications since once might not see a difference before you scale to a certain amount of subscribers / events withins `pubSubBatchDuration`.

I'll let you know if I find something useful, appreciate any feedback.
